drivers/mdio/mdio_dwcxgmac.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT snps_dwcxgmac_mdio
+
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/mdio.h>
+#include <zephyr/init.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(snps_dwcxgmac_mdio, CONFIG_MDIO_LOG_LEVEL);
+
+#define CORE_MDIO_SINGLE_COMMAND_ADDRESS_OFST                  0x200
+#define CORE_MDIO_SINGLE_COMMAND_ADDRESS_RA_SET(value)         (((value) << 0) & 0x0000ffff)
+#define CORE_MDIO_SINGLE_COMMAND_ADDRESS_PA_SET(value)         (((value) << 16) & 0x001f0000)
+#define CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_OFST             0x204
+#define CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SBUSY_SET_MSK    0x00400000
+#define CORE_MDIO_CLAUSE_22_PORT_OFST                          0x220
+#define CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SDATA_SET(value) (((value) << 0) & 0x0000ffff)
+#define CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_CMD_SET(value)   (((value) << 16) & 0x00030000)
+#define CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_CR_SET(value)    (((value) << 19) & 0x00380000)
+#define CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_CRS_SET(value)   (((value) << 31) & 0x80000000)
+#define CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SBUSY_SET(value) (((value) << 22) & 0x00400000)
+#define CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SDATA_GET(value) (((value)&0x0000ffff) >> 0)
+
+struct mdio_dwcxgmac_dev_data {
+	struct k_mutex mdio_transfer_lock;
+};
+
+struct mdio_dwcxgmac_dev_config {
+	DEVICE_MMIO_ROM;
+	uint32_t clk_range;
+	bool clk_range_sel;
+};
+
+static inline int mdio_busy_wait(uint32_t reg_addr, uint32_t bit_msk)
+{
+	uint32_t delay_us = 10000;
+	bool ret;
+
+	ret = WAIT_FOR(!(sys_read32(reg_addr) & bit_msk), delay_us, k_msleep(1));
+	if (ret == false) {
+		return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
+static int mdio_transfer(const struct device *dev, uint8_t prtad, uint8_t devad, uint8_t rw,
```

Comment:
I think it is a good idea to share one function for read and write to potentially save code space in the mdio driver. I will take this as an inspiration, maybe it will be good for other drivers too